nntrainer/layers/bn_layer.cpp

```diff
@@ -147,35 +185,49 @@ void BatchNormalizationLayer::calcDerivative(RunLayerContext &context) {
   Tensor &deriv = context.getIncomingDerivative(SINGLE_INOUT_IDX);
   Tensor &dx = context.getOutgoingDerivative(SINGLE_INOUT_IDX);
   Tensor &deviation = context.getTensor(wt_idx[BNParams::deviation]);
+  Tensor &invstd = context.getTensor(wt_idx[BNParams::invstd]);
+
+  Tensor &t_reduced = context.getTensor(wt_idx[BNParams::t_reduced]);
+  Tensor &t_full = context.getTensor(wt_idx[BNParams::t_full_bw]);
+  Tensor cvar = t_reduced;
 
-  int N = 1;
-  const Tensor &input = context.getInput(SINGLE_INOUT_IDX);
-  const TensorDim &in_dim = input.getDim();
-  for (auto &axis : axes_to_reduce) {
-    N *= in_dim.getTensorDim(axis);
+  if (context.getTrainable()) {
+    /**
+     * This implementation depends on the pre-calculated dbeta calculated.
+     */
+    Tensor &dbeta = context.getWeightGrad(wt_idx[BNParams::beta]);
+    t_reduced = dbeta.divide(divider);
+  } else {
+    t_reduced = deriv.average(axes_to_reduce);
   }
 
-  Tensor dx_1 = gamma.multiply(invstd);
-  Tensor dx_2 = deriv.multiply(N);
-  dx_2.subtract_i(deriv.sum(axes_to_reduce));
-  dx_2.subtract_i(deviation.divide(cvar).multiply(
-    deviation.multiply(deriv).sum(axes_to_reduce)));
+  dx = deriv.subtract(t_reduced);
 
-  dx = dx_2.multiply(dx_1, dx);
-  dx.divide_i(N);
+  t_full = deviation.multiply(deriv);
+  t_full.average(axes_to_reduce, t_reduced);
+  t_reduced.divide_i(cvar);
+  deviation.multiply_i(t_reduced);
+  dx.subtract_i(deviation);
+
+  if (context.getTrainable()) {
+    /**
+     * This calculates dgamma tensor.
+     */
+    Tensor &dgamma = context.getWeightGrad(wt_idx[BNParams::gamma]);
+    t_full.multiply_i(invstd);
+    t_full.sum(axes_to_reduce, dgamma);
+  }
+
+  invstd.multiply_i(gamma);
+  dx.multiply_i(invstd);
 }
 
 void BatchNormalizationLayer::calcGradient(RunLayerContext &context) {
 
-  Tensor &dgamma = context.getWeightGrad(wt_idx[BNParams::gamma]);
   Tensor &dbeta = context.getWeightGrad(wt_idx[BNParams::beta]);
   Tensor &deriv = context.getIncomingDerivative(SINGLE_INOUT_IDX);
-  Tensor &deviation = context.getTensor(wt_idx[BNParams::deviation]);
 
-  dbeta = deriv.sum(axes_to_reduce);
-  Tensor dev = deviation.multiply(invstd);
-  dev.multiply_i(deriv);
-  dgamma = dev.sum(axes_to_reduce);
+  deriv.sum(axes_to_reduce, dbeta);
```

Comment:
only beta is calculated here and gamma is calculated in calcDeriv?

Reply:
yes.